packages/core/src/indexing-store/historical.ts

```diff
@@ -1,363 +1,893 @@
+        }
+
+        sanitizeRecord(record, tableName);
 
-        return decodeRow(row, table, kind);
+        const bytes = getRecordSize(record);
+
+        cacheEntry.record = record;
+        cacheEntry.opIndex = totalCacheOps++;
+        cacheEntry.bytes = bytes;
+
+        return structuredClone(record);
+      }
+    },
+    delete: async ({
+      tableName,
+      id,
+    }: {
+      tableName: string;
+      id: UserId;
+    }) => {
+      const encodedId = getCacheId(id, tableName);
+
+      const insertEntry = storeCache[tableName].insert[encodedId];
+      const updateEntry = storeCache[tableName].update[encodedId];
+      const findEntry = storeCache[tableName].find[encodedId];
+
+      if (insertEntry !== undefined) {
+        const bytes = insertEntry.bytes;
+        delete storeCache[tableName].insert[encodedId];
+
+        cacheSize--;
+        cacheSizeBytes -= bytes;
+
+        return true;
+      } else if (isCacheFull) {
+        return false;
+      } else if (findEntry !== undefined) {
+        if (findEntry.record === undefined) {
+          return false;
+        }
+
+        const bytes = findEntry.bytes;
+        delete storeCache[tableName].find[encodedId];
+
+        cacheSize--;
+        cacheSizeBytes -= bytes;
+      } else if (updateEntry !== undefined) {
+        const bytes = updateEntry.bytes;
+        delete storeCache[tableName].update[encodedId];
+
+        cacheSize--;
+        cacheSizeBytes -= bytes;
       }
-    });
-  },
-  delete: async ({
-    tableName,
-    id,
-  }: {
-    tableName: string;
-    id: UserId;
-  }) => {
-    const table = (schema[tableName] as { table: Table }).table;
-
-    return db.wrap({ method: `${tableName}.delete` }, async () => {
-      const encodedId = encodeValue(id, table.id, kind);
+
+      const table = (schema[tableName] as { table: Table }).table;
 
       const deletedRow = await db
         .withSchema(namespaceInfo.userNamespace)
         .deleteFrom(tableName)
-        .where("id", "=", encodedId)
+        .where(
+          "id",
+          "=",
+          encodeField({ value: id, column: table.id, encoding }),
+        )
         .returning(["id"])
         .executeTakeFirst()
         .catch((err) => {
           throw parseStoreError(err, { id });
         });
 
       return !!deletedRow;
-    });
-  },
-});
+    },
+    flush,
+  };
+};
+
+const getRecordSize = (record: UserRecord) => {
+  // size of metadata
+  let size = 16;
+  for (const col of Object.values(record)) {
+    if (typeof col === "number") {
+      // p.float, p.int
+      size += 8;
+    } else if (typeof col === "string") {
+      // p.hex, p.string, p.enum
+      size += 2 * col.length;
+    } else if (typeof col === "boolean") {
+      // p.boolean
+      size += 4;
+    } else if (typeof col === "bigint") {
+      // p.bigint
+      size += 48;
+    } else if (Array.isArray(col)) {
+      // p.list
+      for (const e of col) {
+        size += getRecordSize(e);
```

Comment:
Doesn't this assume that `col` is an array of objects, because it will hit line 865 above? What happens if it's an array of primitives? Same story for the `p.json()` branch.